gossip/gossip.go

```diff
@@ -0,0 +1,117 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package gossip
+
+import (
+	"sync"
+	"time"
+
+	"github.com/ava-labs/avalanchego/codec"
+	"github.com/ethereum/go-ethereum/log"
+
+	"github.com/ava-labs/coreth/peer"
+	"github.com/ava-labs/coreth/plugin/evm/message"
+)
+
+type Gossiper[T Tx] interface {
+	Mempool() Mempool[T]
+	GossipRequest() message.MempoolTxsRequest
+	Tx() T
+}
+
+func NewPullGossiper[T Tx](
+	gossiper Gossiper[T],
+	networkClient peer.NetworkClient,
+	codec codec.Manager,
+	gossipFrequency time.Duration,
+	gossipSize int,
+	shutdown chan struct{},
+	shutdownWg *sync.WaitGroup,
+) *PullGossiper[T] {
+	return &PullGossiper[T]{
+		gossiper:        gossiper,
+		client:          networkClient,
+		codec:           codec,
+		gossipFrequency: gossipFrequency,
+		gossipSize:      gossipSize,
+		shutdown:        shutdown,
+		shutdownWg:      shutdownWg,
+	}
+}
+
+type PullGossiper[T Tx] struct {
+	gossiper        Gossiper[T]
+	client          peer.NetworkClient
+	codec           codec.Manager
+	gossipFrequency time.Duration
+	gossipSize      int
+	shutdown        chan struct{}
+	shutdownWg      *sync.WaitGroup
+}
+
+// TODO propagate fatal events instead of just logging them
```

Comment:
I think we can do this as a separate pr since a lot of spots in coreth that are in a vm goroutine just log events that could definitely be treated as fatal.